File: bongf/java/src/ch15_binary_search_questions/Q29_installRouter.java
package ch15_binary_search_questions;

import java.util.Arrays;
import java.util.Scanner;

public class Q29_installRouter {
    public static void main(String[] args) {
        Q29_installRouter ir = new Q29_installRouter();
        System.out.println(ir.solution());
    }

    private int solution() {
        Scanner sc = new Scanner(System.in);
        int n = sc.nextInt();
        int c = sc.nextInt();
        int[] arr = new int[n];
        for (int i = 0; i < n; i++) {
            arr[i] = sc.nextInt();
        }
        Arrays.sort(arr);
        int start = 1;
        int end = arr[n-1] - arr[0];
        int result = 0;
        while (start <= end) {
            int mid = (start + end) / 2;
            int cnt = 1;
            int prev = arr[0];
            for(int i =0; i<n; i++) {
                if(arr[i] - prev >= mid) {
                    cnt++;
                    prev = arr[i];
                }
            }
            if(cnt >= c) {
                start = mid + 1;
                result = mid;
            } else {
                end = mid -1;
            }
        }
        return result;
    }
}

File: bongf/python/dongbinbook/ch15_binary_search_questions/q29_install_router.py
# https://www.acmicpc.net/problem/2110
def solution():
    n, c = map(int, input().split())
    homes = []
    for _ in range(n):
        homes.append(int(input()))
    homes.sort()

    start = 1
    end = homes[n-1] - homes[0]
    result = 0
    while start<= end :
        mid = (start+end) //2
        cnt = 1
        prev = homes[0]
        for i in range(1, n):
            if homes[i] - prev >= mid :
                cnt += 1
                prev = homes[i]
        if cnt >= c :
            start = mid + 1
            result = mid
        else:
            end = mid - 1
    return result

print(solution())
